hyperspy/learn/mva.py

```diff
@@ -1636,7 +1642,747 @@ def undo_treatments(self):
+                         algorithm='kmeans',
+                         return_info=False,
+                         **kwargs):
+        """
+        Cluster analysis of a signal or decomposition results of a signal
+
+        Results are stored in `learning_results`.
+
+        Parameters
+        ----------
+        n_clusters : int
+            Number of clusters to find.
+        scaling : {"standard","norm","minmax",None or scikit learn scaling method}
+            default: 'norm'
+            Preprocessing the data before cluster analysis requires scaling
+            the data to be clustered to similar scales. Standard scaling
+            adjusts each feature to have uniform variation. Norm scaling
+            adjusts treats the set of features like a vector and
+            each measurement is scaled to length 1.
+            You can also pass one of the scikit-learn preprocessing
+            scale_method = import sklearn.processing.StandadScaler
+            scaling = scale_method
+            See scaling methods in scikit-learn preprocessing for further
+            details.
+        scaling_kwargs : dict
+            Additional parameters passed to the cluster scaling algorithm.
+            See sklearn.preprocessing scaling methods for further details
+        use_decomposition_results : {True,False,"decomposition","bss"}
+            If True or "decomposition" the signal's decomposition results are used
+            for clustering. If "bss" the blind source separation results are used
+            If False the raw data is used Note that this can be memory intensive 
+            and is only recommened if the Signal has a small signal
+            or navigation dimensions.
+        use_decomposition_results_for_centers : {True,False,"decomposition","bss"}
+            If True or "decomposition" the signal's decomposition results are used
+            for building he cluster centers from the clustered label results.
+            If "bss" the blind source separation results are used.
+            If False the original signal data is used.
+        number_of_components : int, default None
+            If you are getting the cluster centers using the decomposition
+            results (use_decomposition_results_for_centers=True) you can define how
+            many components to use.  If set to None the method uses the
+            estimate of significant components found in the decomposition step
+            using the elbow method and stored in the
+            ``learning_results.number_significant_components`` attribute.
+            This applies to both bss and decomposition results.
+        navigation_mask : boolean numpy array
+            The navigation locations marked as True are not used in the
+            decomposition.
+        signal_mask : boolean numpy array
+            The signal locations marked as True are not used in the
+            decomposition.
+        algorithm : { "kmeans" | "agglomerative" }
```

Comment:
In line with the feature implemented in `decomposition` by @tjof2, would it be possible to pass any sklearn-like clustering algorithm?

Reply:
@tjof needs the components attribute, I need the labels attribute.
I also need n_clusters argument-   MiniBatchKMeans, FeatureAgglomeration, SpectralClustering all take n_clusters and return labels_ so could be added in terms of keyword choices ?

Reply:
> needs the components attribute, I need the labels attribute
I don't think that stops you implementing something _similar_ does it?

Reply:
Recall that you can do:
```
from sklearn.decomposition import PCA`

s.decomposition(algorithm=PCA(n_components=3))
```
So all the arguments are set there.